In [15]:
import pandas as pd

# Load the dataset
data = pd.read_csv("https://raw.githubusercontent.com/YBI-Foundation/Dataset/main/Diabetes%20Missing%20Data.csv")

# Get the missing data counts
missing_data_counts = data.isnull().sum()
print("\nMissing Data Counts:")
print(missing_data_counts)


Missing Data Counts:
Pregnant               0
Glucose                5
Diastolic_BP          35
Skin_Fold            227
Serum_Insulin        374
BMI                   11
Diabetes_Pedigree      0
Age                    0
Class                  0
dtype: int64


In [16]:
print(data.columns)
print(data.shape)  # This will print (number_of_rows, number_of_columns)

Index(['Pregnant', 'Glucose', 'Diastolic_BP', 'Skin_Fold', 'Serum_Insulin',
       'BMI', 'Diabetes_Pedigree', 'Age', 'Class'],
      dtype='object')
(768, 9)


2.2 A variable is a changeable/flexible element that represents some part of a dataset. An observation can be considered to be gathered information from tests or other mediums which contain data that would be collected.

In [29]:
# Assuming you have already loaded your dataset into a DataFrame named df
# For demonstration, let's create a sample DataFrame

data = pd.DataFrame(data)


# Get statistical summary of the numerical columns
summary = df.describe()
print(summary)

         survived      pclass       sibsp       parch        fare
count  891.000000  891.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642    0.523008    0.381594   32.204208
std      0.486592    0.836071    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.000000    0.000000    0.000000
25%      0.000000    2.000000    0.000000    0.000000    7.910400
50%      0.000000    3.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000    8.000000    6.000000  512.329200


In [64]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
print(df.shape)

del df['age']
#del df['deck']

missing_data_indices = df[df.isna().any(axis=1)].index
print(missing_data_indices)

df.drop(index=[61, 829], inplace=True)

print(df.isna().sum())
print(df.describe())


(891, 15)
Index([  0,   2,   4,   5,   7,   8,   9,  12,  13,  14,
       ...
       878, 880, 881, 882, 883, 884, 885, 886, 888, 890],
      dtype='int64', length=690)
survived         0
pclass           0
sex              0
sibsp            0
parch            0
fare             0
embarked         0
class            0
who              0
adult_male       0
deck           688
embark_town      0
alive            0
alone            0
dtype: int64
         survived      pclass       sibsp       parch        fare
count  889.000000  889.000000  889.000000  889.000000  889.000000
mean     0.382452    2.311586    0.524184    0.382452   32.096681
std      0.486260    0.834700    1.103705    0.806761   49.697504
min      0.000000    1.000000    0.000000    0.000000    0.000000
25%      0.000000    2.000000    0.000000    0.000000    7.895800
50%      0.000000    3.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000    1.000000    0.000000   31.000000
max      1.000000    3.0

In the code above, when finding the dimensions of the dataset with df.shape, we find that there are 891 rows and 15 columns. However, when using the df.describe function to try and summarize the code, we get the class "age" to only have  714 rows and see only  6 columns out of 15. Firstly, ages row count is less tthan tthe shape of the dataset becasue itt has missing values. We can find with the function df.isna().sum() that age has 177  missing values, which is the exact difference between the number of rows and age row count, further proving how the age row count discrepancy is caused by its lacking values. Secondly, the reason why df.describe shows fewer columns then what df.shape says there are because there are  9 columns in the dataset that are non-numerical. Since df.describe only looks at numerical values,  then these non-numerical columns are not shown when using this function.

An attribute is a variable that only states and stores information about an object to which it is also attached. A key characteristic of attributes is that they are called upon without the use of parenthesis. This contrasts an attribute as a method or "function" using parenthesis. Furthermore, methods don't state or keep data from an object but instead perform computations or operations.

Count: The number of non-missing values in each column
Mean: The average of all the values in a column
Std: The measure of the deviation over a data set, based on the mean
Min: The minimum value in each column
Max: The maximum value in each column
25%: The twenty-fifth percentile of the data. Also known as the first quartile, denoted Q1
50%: The fiftieth percentile of the data. Also known as the median
75%: The seventy-fifth percentile of the data. Also known as the third quartile, denoted Q3

A scenario where it would be beneficial to use df.dropna() over del df['col'] is when a dataset has more missing data in rows than columns. For instance, a dataset where the rows contain the ages of people and the columns contain headings of things only 18-year-olds can do, then the rows of people aged 1-17 will be empty. If we were to remove the rows containing missing values, then all rows would be removed as every column contains a row(s) of ages that are not 18+, thus being empty. However, by removing rows containing nothing, then only 17 completely empty rows will be removed, and no data will be lost.

A scenario where it would be beneficial to use del df['col'] over df.dropna() is when a dataset has more missing data in a column than rows. For instance, a dataset that would benefit from doing just this is a medical dataset where the columns are comprised of different people, and the rows contain various ailments. If the majority of columns are empty (people don't have certain aliments), then it would be best to remove those columns over the rows containing the missing data because a single person will remove many missing data spots, whereas a row will remove many filled data spots, so to increase data retention df['col'] should be used over df.dropna().

It could be important to do del df['col'] before df.dropna() when both are going to be used because it decreases the percentage of data lost. Imagine a data set with ten rows and columns with ten missing data points. All but one missing data point are in one column, but the other ones are dispersed in different columns and rows. There are two cases to look at in order to conclude if, in this scenario, we should use del df['col'] first.

Case 1: We use df.dropna() then del df['col']

By using df.dropna() first, since all rows contain 1 missing  data point then by removing a row, 9% of real data  will be llostt.

Case 1: We use del df['col'] then df.dropna()
By using del df['col'] first, since 90% of missing data is in one column, tthen by removing thatt column youd only be losing 1% of real data, tthen to remove the final missing data youd have to lose a row (9% of real data). 

Therefore, by using df.dropna() first you can get rid of 100% of missing data points at the cost of 10% of real data, whereas if you choose the former, you will remove 9% of real data for only 1% of the missing data. So in this case, it is a no-brainer to use df.dropna() first.

For the data set I used in question 4, for me to remove all missing data, I would remove the columns labelled "age," and "deck" with del df['col'] then I would remove the rows labelled "embark_town" and "embark" using df. drop (). I would remove the missing data to increase data retention. I am using df.isna().sum() helps me see that the majority of missing data is in 2 columns, and using df[df.isna().any(axis=1)].index, I can find which rows have the remaining missing data. Of the 869 missing data, 865 are in 2 columns, and 4 are in two other rows. This results in 903 real data being lost. If I we're to remove the rows and columns containing missing data first, I'd remove over 10000 real data to remove the missing data, so, of course, I'd choose the first method to remove all the missing data as it contains more data.

In [14]:
import pandas as pd

# Load Titanic dataset
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

# Example: Group by 'Sex' and describe 'Age'
result = df.groupby("Sex")["Age"].describe()
print(result)
print(df.describe())

        count       mean        std   min   25%   50%   75%   max
Sex                                                              
female  261.0  27.915709  14.110146  0.75  18.0  27.0  37.0  63.0
male    453.0  30.726645  14.678201  0.42  21.0  29.0  39.0  80.0
       PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      

df.groupby("col1")["col2"].describe() does a few things. ("col1") takes the column listed and makes several classes or groups with the number of unique elements it has. ["col2"] takes the column listed and puts the column data into one of the classes created by ("col1"), depending on each data point corresponding to the row value in ("col1"). So, using the Titanic dataset ("Age") creates two classes labelled "female" or "male" as those are the unique values in "Sex." So ["Age"] will put its data into either class depending on each of its data points corresponding to the "Sex" data point in that row. 

The values in count differ between df.describe() and df.group by ("Sex")["Age"].describe() because using df. group by ("col1")["col2"].describe() gives a count within subgroups, whereas df.describe() takes into account all non-null data for the whole data set not just certain groups.

A chatbot is much more useful than Google Search for my problems. This is because I don't understand Python very well, and when I search for my errors on Google, I get results that do answer my question, but since I don't understand the terminology used, I get lost or intimidated and never actually understand what is wrong and how to fix it. Contrastly, a chatbot is much more effective for a person like me as I can converse with it in person, and it will output answers that are comprehensive to someone of my calibre of coding knowledge with answers straight to the point. I'll almost always go to a chatbot for help before googling.

Kinda. I read through some of the course wiki-textbook, but mainly relied on a chatbot to help me.

Chatbot 1 Summary of Interaction:
In this session, I sought assistance on various data analysis concepts and methods related to a dataset. The conversation covered the following topics:

Loading and Inspecting a Dataset:

I was provided with steps to load a dataset from a URL using Python’s pandas library and shown how to inspect its structure, including viewing the columns (df.columns), getting summary statistics (df.describe()), and checking the shape of the data (df.shape).
Definitions of Observations and Variables:

I learned the definitions of key terms:
Observations: The individual entries or records in a dataset (e.g., each character in an Animal Crossing dataset).
Variables: The attributes or properties of each observation (e.g., Name, Species, Birthday).
Using df.describe() and df['column'].value_counts() Methods:

I was given examples of using these methods:
df.describe(): Provides a summary of key statistics (count, mean, standard deviation, min, max, and percentiles) for numerical columns.
df['column'].value_counts(): Counts the occurrences of each unique value in a column, typically used for categorical or discrete numerical data.
Difference Between Attributes and Methods:

I was explained the difference between:
Attributes: Variables that belong to an object (e.g., df.shape), accessed without parentheses.
Methods: Functions that perform actions or computations (e.g., df.describe()), requiring parentheses.
Explanation of Summary Statistics from df.describe():

A detailed explanation of the following summary statistics was provided: count, mean, standard deviation (std), minimum (min), 25th percentile (25%), median (50%), 75th percentile (75%), and maximum (max). Definitions and examples were shared for each, helping me understand what they represent in data analysis.
This conversation helped clarify important data analysis concepts, particularly related to inspecting and summarizing a dataset using pandas in Python. The explanations and examples given will be useful for my assignment on exploring and summarizing data in a structured format.

Link: https://chatgpt.com/c/66df97a3-45c8-800c-95f1-a434dc4e941d

Chatbot 2 Summary of Interaction:
In this interaction, we addressed a series of common coding errors encountered while working with the Titanic dataset in Python, using the pandas library. Here's a summary of the issues and how they were resolved:

NameError: name 'pd' is not defined:

Issue: The pandas library wasn’t imported.
Solution: The user was instructed to add import pandas as pd to import the pandas library correctly.
FileNotFoundError: No such file or directory: 'titanic.csv':

Issue: The file 'titanic.csv' could not be found.
Solution: The user was advised to check the file name and provide the correct file path, ensuring it matched the location of the dataset.
NameError: name 'DF' is not defined:

Issue: A variable named DF (uppercase) was used without being defined.
Solution: The user was instructed to either define DF properly or use the lowercase df, which is commonly used after loading the dataset.
SyntaxError: '(' was never closed:

Issue: A missing closing parenthesis caused a syntax error.
Solution: The user was guided to ensure all open parentheses were properly closed.
AttributeError: 'SeriesGroupBy' object has no attribute 'describle':

Issue: A typo in the method .describle() instead of .describe().
Solution: The correct method .describe() was provided.
KeyError: 'age':

Issue: The column name "age" was incorrect due to case sensitivity (the correct name was "Age").
Solution: The user was advised to check column names and correct the case.
KeyError: 'sex':

Issue: The column name "sex" was not found.
Solution: The user was instructed to use the correct column name "Sex" (case-sensitive) and check for any spaces.
NameError: name 'Age' is not defined:

Issue: The column name "Age" was treated as a variable instead of a string.
Solution: The user was advised to enclose column names in quotes (e.g., "Age").
NameError: name 'Sex' is not defined:

Issue: The column name "Sex" was treated as a variable instead of a string.
Solution: The user was advised to enclose "Sex" in quotes to resolve the error.
Overall Learnings:
Case sensitivity matters when referencing column names in pandas ("Age" vs. "age").
Syntax issues like unclosed parentheses or misspelled methods can cause errors.
File paths need to be correct and relative to the working directory.
Always ensure column names are enclosed in quotes when referencing them in pandas.
By addressing each error systematically, we ensured the user’s code would run successfully with the Titanic dataset.

Link: https://chatgpt.com/c/66e235b8-a478-800c-b8aa-b49c80a22b28